Deriving `Clone` and `Copy` twice generates bad diagnostics

### Code
```rust
#[derive(Clone, Copy)]
#[derive(Clone, Copy)]
enum E {}
```

### Current output
```
error[E0119]: conflicting implementations of trait `Clone` for type `E`
 --> src/lib.rs:2:10
  |
1 | #[derive(Clone, Copy)]
  |          ----- first implementation here
2 | #[derive(Clone, Copy)]
  |          ^^^^^ conflicting implementation for `E`
  |
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Copy` for type `E`
 --> src/lib.rs:2:17
  |
1 | #[derive(Clone, Copy)]
  |                 ---- first implementation here
2 | #[derive(Clone, Copy)]
  |                 ^^^^ conflicting implementation for `E`
  |
  = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `*self` which is behind a shared reference
 --> src/lib.rs:2:10
  |
2 | #[derive(Clone, Copy)]
  |          ^^^^^ move occurs because `*self` has type `E`, which does not implement the `Copy` trait
  |
note: if `E` implemented `Clone`, you could clone the value
 --> src/lib.rs:3:1
  |
2 | #[derive(Clone, Copy)]
  |          ----- you could clone this value
3 | enum E {}
  | ^^^^^^ consider implementing `Clone` for this type
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `*self` which is behind a shared reference
 --> src/lib.rs:1:10
  |
1 | #[derive(Clone, Copy)]
  |          ^^^^^ move occurs because `*self` has type `E`, which does not implement the `Copy` trait
  |
note: if `E` implemented `Clone`, you could clone the value
 --> src/lib.rs:3:1
  |
1 | #[derive(Clone, Copy)]
  |          ----- you could clone this value
2 | #[derive(Clone, Copy)]
3 | enum E {}
  | ^^^^^^ consider implementing `Clone` for this type
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Desired output
```
error[E0119]: conflicting implementations of trait `Clone` for type `E`
 --> src/lib.rs:2:10
  |
1 | #[derive(Clone, Copy)]
  |          ----- first implementation here
2 | #[derive(Clone, Copy)]
  |          ^^^^^ conflicting implementation for `E`
  |
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Copy` for type `E`
 --> src/lib.rs:2:17
  |
1 | #[derive(Clone, Copy)]
  |                 ---- first implementation here
2 | #[derive(Clone, Copy)]
  |                 ^^^^ conflicting implementation for `E`
  |
  = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)
```
### Rationale and extra context

First encountered this with a macro that automatically derived `Clone` and `Copy`. The later two errors are pointing to internal implementation details and have wrong spans; they should be removed. They also suggest "consider implementing `Clone` for this type", which is nonsense.

### Other cases

_No response_

### Rust Version

1.83.0 nightly on playground

### Anything else?

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"VulnBandit"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_borrowck/src/diagnostics/move_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/move_errors.rs
@@ -10,6 +10,7 @@ use rustc_middle::ty::{self, Ty};
 use rustc_mir_dataflow::move_paths::{LookupResult, MovePathIndex};
 use rustc_span::{BytePos, ExpnKind, MacroKind, Span};
 use rustc_trait_selection::error_reporting::traits::FindExprBySpan;
+use rustc_trait_selection::infer::InferCtxtExt;
 use tracing::debug;
 
 use crate::MirBorrowckCtxt;
@@ -267,6 +268,15 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, 'infcx, 'tcx> {
                 kind,
                 self.is_upvar_field_projection(original_path.as_ref())
             );
+            if self.has_ambiguous_copy(original_path.ty(self.body, self.infcx.tcx).ty) {
+                // If the type may implement Copy, skip the error.
+                // It's an error with the Copy implementation (e.g. duplicate Copy) rather than borrow check
+                self.dcx().span_delayed_bug(
+                    span,
+                    "Type may implement copy, but there is no other error.",
+                );
+                return;
+            }
             (
                 match kind {
                     &IllegalMoveOriginKind::BorrowedContent { target_place } => self
@@ -289,6 +299,13 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, 'infcx, 'tcx> {
 
         self.add_move_hints(error, &mut err, err_span);
         self.buffer_error(err);
+    }
+
+    fn has_ambiguous_copy(&mut self, ty: Ty<'tcx>) -> bool {
+        let Some(copy_trait_def) = self.infcx.tcx.lang_items().copy_trait() else { return false };
+        // This is only going to be ambiguous if there are incoherent impls, because otherwise
+        // ambiguity should never happen in MIR.
+        self.infcx.type_implements_trait(copy_trait_def, [ty], self.param_env).may_apply()
     }
 
     fn report_cannot_move_from_static(&mut self, place: Place<'tcx>, span: Span) -> Diag<'infcx> {

--- a/tests/ui/derives/duplicate-derive-copy-clone-diagnostics.rs
+++ b/tests/ui/derives/duplicate-derive-copy-clone-diagnostics.rs
@@ -1,0 +1,11 @@
+// Duplicate implementations of Copy/Clone should not trigger
+// borrow check warnings
+// See #131083
+
+#[derive(Copy, Clone)]
+#[derive(Copy, Clone)]
+//~^ ERROR conflicting implementations of trait `Clone` for type `E`
+//~| ERROR conflicting implementations of trait `Copy` for type `E`
+enum E {}
+
+fn main() {}

--- a/tests/ui/derives/duplicate-derive-copy-clone-diagnostics.stderr
+++ b/tests/ui/derives/duplicate-derive-copy-clone-diagnostics.stderr
@@ -1,0 +1,23 @@
+error[E0119]: conflicting implementations of trait `Copy` for type `E`
+  --> $DIR/duplicate-derive-copy-clone-diagnostics.rs:6:10
+   |
+LL | #[derive(Copy, Clone)]
+   |          ---- first implementation here
+LL | #[derive(Copy, Clone)]
+   |          ^^^^ conflicting implementation for `E`
+   |
+   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0119]: conflicting implementations of trait `Clone` for type `E`
+  --> $DIR/duplicate-derive-copy-clone-diagnostics.rs:6:16
+   |
+LL | #[derive(Copy, Clone)]
+   |                ----- first implementation here
+LL | #[derive(Copy, Clone)]
+   |                ^^^^^ conflicting implementation for `E`
+   |
+   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0119`.

